### DISCUSSION TITLE
how to know i have connect to server.

### DISCUSSION THREAD

**Discussion by LucasMou** [edenhill/librdkafka#4445] [Q&A]
I want to know the status of the connection to the server. I currently check whether the client has established a connection to the server through the rd_kafka_metadata interface, but I don't think this is the best solution. I think there must be a better implementation for kafka to notify the application layer of connection and disconnection events.

**Comment by rolandyoung** [comment]
Please see this FAQ article https://github.com/confluentinc/librdkafka/wiki/FAQ#i-want-to-get-an-event-when-brokers-come-up-and-down (and possibly, the following one). It turns out that "the status of the connection" is not as well-defined as you might think.
In general, if a subscriber can read messages successfully, the connection is up and good; if it gets an error when trying to do so, the connection is not good, but librdkafka is probably trying to reestablish it; otherwise, it can best be regarded as being in an idle state. Same for publishing, but, in this case, you do not know that the connection is good until you get a delivery receipt.
